masonry/src/passes/layout.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! The layout pass, which provides the size and position of each widget
+//! before any translations applied in [`compose`](crate::passes::compose).
+//! Most of the logic for this pass happens in [`Widget::layout`] implementations.
+
+use tracing::{info_span, trace, warn};
+use vello::kurbo::{Point, Rect, Size};
+
+use crate::render_root::RenderRoot;
+use crate::widget::WidgetState;
+use crate::{BoxConstraints, LayoutCtx, Widget, WidgetPod};
+
+// TODO - negative rects?
+/// Return `true` if all of `smaller` is within `larger`.
+fn rect_contains(larger: &Rect, smaller: &Rect) -> bool {
+    smaller.x0 >= larger.x0
+        && smaller.x1 <= larger.x1
+        && smaller.y0 >= larger.y0
+        && smaller.y1 <= larger.y1
+}
+
+pub(crate) fn run_layout_inner<W: Widget>(
+    parent_ctx: &mut LayoutCtx<'_>,
+    pod: &mut WidgetPod<W>,
+    bc: &BoxConstraints,
+) -> bool {
+    let id = pod.id().to_raw();
+    let widget_mut = parent_ctx
+        .widget_children
+        .get_child_mut(id)
+        .expect("WidgetPod: inner widget not found in widget tree");
+    let mut state_mut = parent_ctx
+        .widget_state_children
+        .get_child_mut(id)
+        .expect("WidgetPod: inner widget not found in widget tree");
+    let widget = widget_mut.item;
+    let state = state_mut.item;
+
+    if state.is_stashed {
+        debug_panic!(
+            "Error in '{}' #{}: trying to compute layout of stashed widget.",
+            widget.short_type_name(),
+            id,
+        );
+        state.size = Size::ZERO;
+        return false;
+    }
+
+    state.needs_compose = true;
+    state.is_expecting_place_child_call = true;
+    // TODO - Not everything that has been re-laid out needs to be repainted.
+    state.needs_paint = true;
+    state.request_accessibility = true;
+    state.needs_accessibility = true;
+
+    bc.debug_check(widget.short_type_name());
+    trace!("Computing layout with constraints {:?}", bc);
+
+    state.local_paint_rect = Rect::ZERO;
+
+    let new_size = {
+        let mut inner_ctx = LayoutCtx {
+            widget_state: state,
+            widget_state_children: state_mut.children.reborrow_mut(),
+            widget_children: widget_mut.children,
+            global_state: parent_ctx.global_state,
+            mouse_pos: parent_ctx.mouse_pos,
+        };
+
+        widget.layout(&mut inner_ctx, bc)
+    };
+
+    // TODO - One we add request_layout, check for that flag.
+    // If it's true, that's probably an error.
+    state.needs_layout = false;
+
+    state.local_paint_rect = state
+        .local_paint_rect
+        .union(new_size.to_rect() + state.paint_insets);
```

Comment:
I'm not a *huge* fan of using `Add` for `Insets`, but that's a Kurbo thing - this is an idiomatic usage thereof